File: _includes/code/java-v6/src/test/java/_ManageCollectionsMultiTenancyTest.java
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.collections.CollectionHandle;
import io.weaviate.client6.v1.api.collections.tenants.Tenant;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

class _ManageCollectionsMultiTenancyTest {

  private static WeaviateClient client;

  @BeforeAll
  public static void beforeAll() {
    String openaiApiKey = System.getenv("OPENAI_API_KEY");
    assertThat(openaiApiKey).isNotBlank()
        .withFailMessage("Please set the OPENAI_API_KEY environment variable.");

    client = WeaviateClient.connectToLocal(config -> config
        .setHeaders(Map.of("X-OpenAI-Api-Key", openaiApiKey)));
  }

  @AfterEach
  public void afterAll() throws Exception {
    client.close();
  }

  //TODO[g.-despot] A lot of strange errors: IllegalState Connection pool shut down
  @Test
  void testEnableMultiTenancy() throws IOException {
    // START EnableMultiTenancy
    client.collections.create("MultiTenancyCollection", col -> col
        .multiTenancy(mt -> mt.enabled(true)));
    // END EnableMultiTenancy

    var config = client.collections.getConfig("MultiTenancyCollection").get();
    assertThat(config.multiTenancy().createAutomatically()).isTrue();
  }

  @Test
  void testEnableAutoActivationMultiTenancy() throws IOException {
    // START EnableAutoActivation
    client.collections.create("MultiTenancyCollection", col -> col
        .multiTenancy(mt -> mt.autoTenantActivation(true)));
    // END EnableAutoActivation

    var config = client.collections.getConfig("MultiTenancyCollection").get();
    assertThat(config.multiTenancy().activateAutomatically()).isTrue();
  }

  @Test
  void testEnableAutoMT() throws IOException {
    // START EnableAutoMT
    client.collections.create("CollectionWithAutoMTEnabled", col -> col
        .multiTenancy(mt -> mt
            .autoTenantCreation(true)));
    // END EnableAutoMT

    var config = client.collections.getConfig("CollectionWithAutoMTEnabled").get();
    assertThat(config.multiTenancy().createAutomatically()).isTrue();
  }

  @Test
  void testUpdateAutoMT() throws IOException {
    String collectionName = "MTCollectionNoAutoMT";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt
            .autoTenantActivation(false)));

    // START UpdateAutoMT
    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.config.update(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));
    // END UpdateAutoMT

    var config = client.collections.getConfig(collectionName).get();
    assertThat(config.multiTenancy().createAutomatically()).isTrue();
  }

  @Test
  void testAddTenantsToClass() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);

    // START AddTenantsToClass
    collection.tenants.create(
        Tenant.active("tenantA"),
        Tenant.active("tenantB"));
    // END AddTenantsToClass

    List<Tenant> tenants = collection.tenants.get();
    assertThat(tenants).hasSize(2);
    assertThat(tenants.get(0).name())
        .containsAnyOf("tenantA", "tenantB");
  }

  @Test
  void testListTenants() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.tenants.create(
        Tenant.active("tenantA"),
        Tenant.active("tenantB"));

    // START ListTenants
    List<Tenant> tenants = collection.tenants.get();
    System.out.println(tenants);
    // END ListTenants

    assertThat(tenants).hasSize(2);
  }

  @Test
  void testGetTenantsByName() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.tenants.create(
        Tenant.active("tenantA"),
        Tenant.active("tenantB"));

    // // START GetTenantsByName
    List<String> tenantNames = Arrays.asList("tenantA", "tenantB",
        "nonExistentTenant");
    List<Tenant> tenants = collection.tenants.get(tenantNames);
    System.out.println(tenants);
    // // END GetTenantsByName

    assertThat(tenants).hasSize(2);
  }

  @Test
  void testGetOneTenant() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.tenants.create(Tenant.active("tenantA"));

    // // START GetOneTenant
    String tenantName = "tenantA";
    Optional<Tenant> tenant = collection.tenants.get(tenantName);
    System.out.println(tenant);
    // // END GetOneTenant

    assertThat(tenant).isPresent();
  }

  @Test
  void testActivateTenant() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    // // START ActivateTenants
    String tenantName = "tenantA";
    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.tenants.activate(tenantName);
    // // END ActivateTenants

    Optional<Tenant> tenant = collection.tenants.get(tenantName);
    assertThat(tenant).isPresent();
  }

  @Test
  void testDeactivateTenant() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    // // START DeactivateTenants
    String tenantName = "tenantA";
    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.tenants.deactivate(tenantName);
    // // END DeactivateTenants

    Optional<Tenant> tenant = collection.tenants.get(tenantName);
    assertThat(tenant).isPresent();
  }

  @Test
  void testOffloadTenant() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    // // START OffloadTenants
    String tenantName = "tenantA";
    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.tenants.offload(tenantName);
    // // END OffloadTenants

    Optional<Tenant> tenant = collection.tenants.get(tenantName);
    assertThat(tenant).isPresent();
  }

  @Test
  void testRemoveTenants() throws IOException {
    String collectionName = "MultiTenancyCollection";
    client.collections.create(collectionName, col -> col
        .multiTenancy(mt -> mt.autoTenantCreation(true)));

    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);
    collection.tenants.create(
        Tenant.active("tenantA"),
        Tenant.active("tenantB"));

    // // START RemoveTenants
    collection.tenants.delete(Arrays.asList("tenantB", "tenantX"));
    // // END RemoveTenants

    List<Tenant> tenants = collection.tenants.get();
    assertThat(tenants).hasSize(1);
    assertThat(tenants.get(0).name())
        .containsAnyOf("tenantA");
  }
}

File: _includes/code/csharp/_ManageCollectionsMultiTenancyTest.cs
using Xunit;
using Weaviate.Client;
using Weaviate.Client.Models;
using System;
using System.Threading.Tasks;
using System.Linq;
using System.Collections.Generic;

namespace WeaviateProject.Tests;

public class ManageCollectionsMultiTenancyTest : IAsyncLifetime
{
    private WeaviateClient client;

    // Runs before each test (like @BeforeEach)
    public Task InitializeAsync()
    {
        string openaiApiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
        if (string.IsNullOrWhiteSpace(openaiApiKey))
        {
            throw new ArgumentException("Please set the OPENAI_API_KEY environment variable.");
        }

        // Note: The C# client doesn't support setting headers like 'X-OpenAI-Api-Key' via the constructor for local connections.
        // This must be configured in Weaviate's environment variables.
        client = new WeaviateClient(new ClientConfiguration { RestAddress = "localhost", RestPort = 8080 });

        return Task.CompletedTask;
    }

    // Runs after each test (like @AfterEach)
    public async Task DisposeAsync()
    {
        // Clean up any collections created during the tests
        await client.Collections.DeleteAll();
    }

    [Fact]
    public async Task TestEnableMultiTenancy()
    {
        // START EnableMultiTenancy
        await client.Collections.Create(new Collection
        {
            Name = "MultiTenancyCollection",
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true }
        });
        // END EnableMultiTenancy

        var config = await client.Collections.Export("MultiTenancyCollection");
        Assert.True(config.MultiTenancyConfig.Enabled);
    }

    [Fact]
    public async Task TestEnableAutoActivationMultiTenancy()
    {
        // START EnableAutoActivation
        await client.Collections.Create(new Collection
        {
            Name = "MultiTenancyCollection",
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true, AutoTenantActivation = true }
        });
        // END EnableAutoActivation

        var config = await client.Collections.Export("MultiTenancyCollection");
        Assert.True(config.MultiTenancyConfig.AutoTenantActivation);
    }

    [Fact]
    public async Task TestEnableAutoMT()
    {
        // START EnableAutoMT
        await client.Collections.Create(new Collection
        {
            Name = "CollectionWithAutoMTEnabled",
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true, AutoTenantCreation = true }
        });
        // END EnableAutoMT

        var config = await client.Collections.Export("CollectionWithAutoMTEnabled");
        Assert.True(config.MultiTenancyConfig.AutoTenantCreation);
    }

    [Fact]
    public async Task TestUpdateAutoMT()
    {
        string collectionName = "MTCollectionNoAutoMT";
        await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true, AutoTenantCreation = false }
        });

        // START UpdateAutoMT
        var collection = client.Collections.Use(collectionName);
        await collection.Config.Update(c =>
        {
            c.MultiTenancyConfig.AutoTenantCreation = true;
        });
        // END UpdateAutoMT

        var config = await client.Collections.Export(collectionName);
        Assert.True(config.MultiTenancyConfig.AutoTenantCreation);
    }

    [Fact]
    public async Task TestAddTenantsToClass()
    {
        string collectionName = "MultiTenancyCollection";
        await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true }
        });

        var collection = client.Collections.Use(collectionName);

        // START AddTenantsToClass
        await collection.Tenants.Add(
            new Tenant { Name = "tenantA" },
            new Tenant { Name = "tenantB" }
        );
        // END AddTenantsToClass

        var tenants = (await collection.Tenants.List()).ToList();
        Assert.Equal(2, tenants.Count);
        Assert.Contains(tenants, t => t.Name == "tenantA");
        Assert.Contains(tenants, t => t.Name == "tenantB");
    }

    [Fact]
    public async Task TestListTenants()
    {
        string collectionName = "MultiTenancyCollection";
        var collection = await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true }
        });
        await collection.Tenants.Add(new Tenant { Name = "tenantA" }, new Tenant { Name = "tenantB" });

        // START ListTenants
        var tenants = await collection.Tenants.List();
        foreach (var t in tenants) Console.WriteLine(t.Name);
        // END ListTenants

        Assert.Equal(2, tenants.Count());
    }

    [Fact]
    public async Task TestGetTenantsByName()
    {
        string collectionName = "MultiTenancyCollection";
        var collection = await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true }
        });
        await collection.Tenants.Add(new Tenant { Name = "tenantA" }, new Tenant { Name = "tenantB" });

        // START GetTenantsByName
        var tenantNames = new[] { "tenantA", "tenantB", "nonExistentTenant" };
        var tenants = await collection.Tenants.List(tenantNames);
        foreach (var t in tenants) Console.WriteLine(t.Name);
        // END GetTenantsByName

        Assert.Equal(2, tenants.Count());
    }

    [Fact]
    public async Task TestGetOneTenant()
    {
        string collectionName = "MultiTenancyCollection";
        var collection = await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true }
        });
        await collection.Tenants.Add(new Tenant { Name = "tenantA" });

        // START GetOneTenant
        string tenantName = "tenantA";
        var tenant = await collection.Tenants.Get(tenantName);
        Console.WriteLine(tenant?.Name);
        // END GetOneTenant

        Assert.NotNull(tenant);
    }

    [Fact]
    public async Task TestActivateTenant()
    {
        string collectionName = "MultiTenancyCollection";
        var collection = await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true }
        });
        await collection.Tenants.Add(new Tenant { Name = "tenantA", Status = TenantActivityStatus.Inactive });

        // START ActivateTenants
        string tenantName = "tenantA";
        await collection.Tenants.Activate(tenantName);
        // END ActivateTenants

        var tenant = await collection.Tenants.Get(tenantName);
        Assert.Equal(TenantActivityStatus.Active, tenant?.Status);
    }

    [Fact]
    public async Task TestDeactivateTenant()
    {
        string collectionName = "MultiTenancyCollection";
        var collection = await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true, AutoTenantCreation = true }
        });
        await collection.Tenants.Add(new Tenant { Name = "tenantA" });

        // START DeactivateTenants
        string tenantName = "tenantA";
        await collection.Tenants.Deactivate(tenantName);
        // END DeactivateTenants

        var tenant = await collection.Tenants.Get(tenantName);
        Assert.Equal(TenantActivityStatus.Inactive, tenant?.Status);
    }

    // START OffloadTenants
    // Note: 'Offload' is not a current concept in the client. Use 'Deactivate' for similar functionality.
    // Coming soon
    // END OffloadTenants

    [Fact]
    public async Task TestRemoveTenants()
    {
        string collectionName = "MultiTenancyCollection";
        var collection = await client.Collections.Create(new Collection
        {
            Name = collectionName,
            MultiTenancyConfig = new MultiTenancyConfig { Enabled = true }
        });
        await collection.Tenants.Add(new Tenant { Name = "tenantA" }, new Tenant { Name = "tenantB" });

        // START RemoveTenants
        await collection.Tenants.Delete(new[] { "tenantB", "tenantX" });
        // END RemoveTenants

        var tenants = (await collection.Tenants.List()).ToList();
        Assert.Single(tenants);
        Assert.Equal("tenantA", tenants.First().Name);
    }
}